xds/internal/client/requests_counter.go

```diff
@@ -56,11 +56,12 @@ func GetServiceRequestsCounter(serviceName string) *ServiceRequestsCounter {
 // StartRequest starts a request for a service, incrementing its number of
 // requests by 1. Returns an error if the max number of requests is exceeded.
 func (c *ServiceRequestsCounter) StartRequest(max uint32) error {
-	if atomic.LoadUint32(&c.numRequests) >= max {
-		return fmt.Errorf("max requests %v exceeded on service %v", max, c.ServiceName)
+	newNum := atomic.AddUint32(&c.numRequests, 1)
+	if newNum <= max {
+		return nil
 	}
-	atomic.AddUint32(&c.numRequests, 1)
-	return nil
+	atomic.AddUint32(&c.numRequests, ^uint32(0))
```

Comment:
Hmm, this will race the other way.

Reply:
Should we instead have a loop:

```
for {
  - curr := load the atomic value
  - if curr >= max, error out
  - compareAndSwap(curr, curr+1)
  - if CAS fails, retry the loop, else return
  }
  ```

Reply:
This becomes almost equivalent to a lock in that case.  I think we are willing to sacrifice precise results for performance since this is the RPC path.  Let's make sure we are consistent with the other implementations, here.